# Profitable apps
This project shows the profitable apps given all the data reclolected from movile apps

In [1]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [2]:
opened_file1 = open("AppleStore.csv", "r", encoding="utf-8")
from csv import reader
readed_file1 = reader(opened_file1)
apple_store_data = list(readed_file1)
#print(apple_store_data[1:5])

opened_file2 = open("googleplaystore.csv", "r", encoding="utf-8")
readed_file2 = reader(opened_file2)
google_store_data = list(readed_file2)
#print(google_store_data[1:5])

In [3]:
#print(explore_data(apple_store_data, 1, 3))

# Ver duplicados
Note como si hay duplicados y los debeme eliminar

In [4]:
for data in google_store_data[1:]:
    name = data[0]
    if name == 'Instagram':
        print(data)
        print('\n')
    

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']




# remove duplicated entries
Compare por ejemplo con los repetidos de instagram y note como si se elimnaron los repetidos y además quedí el de más alto puntaje, es decir el último
- el tamaño del dicccion rio debe dar al final 9659

In [7]:
#del(google_store_data[10473])# ejecutelo solo una vez

In [9]:
def last_updated(data_list, name_index, recently_column):
    reviews_max = {}
    for data in data_list[1:]:
        #print(data)
        name = data[name_index]
        n_reviews = float(data[recently_column])
        if name in reviews_max and reviews_max[name] < n_reviews:        
            reviews_max[name] = n_reviews

        if name not in reviews_max:
            reviews_max[name] = n_reviews
    return reviews_max

android_reviews_max = last_updated(google_store_data,0,3)        
print(android_reviews_max['Instagram'])
print(len(google_store_data))#antes
print(len(android_reviews_max))#despues

"""
# al parecer apple_store_data no requiere esta limpieza
# pues sus datos son únicos
print('IOS')
ios_reviews_max = last_updated(apple_store_data,1) 
print(len(apple_store_data))#antes
print(len(ios_reviews_max))#despues

"""


66577446.0
10841
9659


"\n# al parecer apple_store_data no requiere esta limpieza\n# pues sus datos son únicos\nprint('IOS')\nios_reviews_max = last_updated(apple_store_data,1) \nprint(len(apple_store_data))#antes\nprint(len(ios_reviews_max))#despues\n\n"

In [10]:


android_clean = []
android_already_added = []
for data in google_store_data[1:]:
    name = data[0]
    n_reviews = float(data[3])
    if android_reviews_max[name] ==  n_reviews and name not in android_already_added:
        android_clean.append(data)
        android_already_added.append(name)
        
print('android Con duplicados => ',len(google_store_data))
print('android Sin duplicados => ',len(android_clean))




android Con duplicados =>  10841
android Sin duplicados =>  9659


- Note como hay apps de otro idioma pero el objetivo son solo las apps en inglés, debemos limpiar acá también :) 

In [11]:
print(android_clean[4412])
print(android_clean[7940])
print('\n')
print('Apple apps => ')
print(apple_store_data[814])
print(apple_store_data[6732])

['中国語 AQリスニング', 'FAMILY', 'NaN', '21', '17M', '5,000+', 'Free', '0', 'Everyone', 'Education', 'June 22, 2016', '2.4.0', '4.0 and up']
['لعبة تقدر تربح DZ', 'FAMILY', '4.2', '238', '6.8M', '10,000+', 'Free', '0', 'Everyone', 'Education', 'November 18, 2016', '6.0.0.0', '4.1 and up']


Apple apps => 
['445375097', '爱奇艺PPS -《欢乐颂2》电视剧热播', '224617472', 'USD', '0.0', '14844', '0', '4.0', '0.0', '6.3.3', '17+', 'Entertainment', '38', '5', '3', '1']
['1120021683', '【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜', '77551616', 'USD', '0.0', '0', '0', '0.0', '0.0', '1.3', '12+', 'Games', '38', '0', '1', '1']


-Una forma de identificar el idioma en que lo escribieron puede ser detectando el entero UNICODE que representa el string 

In [12]:
print(ord('a'))
print(ord('Z'))
print(ord('5'))
print(ord('爱'))

97
90
53
29233


- Function that takes in a string and returns False if there's any character in the string that doesn't belong to the set of common English characters, otherwise it returns True

In [13]:
def common_english(str_input):
    for character in str_input:
        #print(ord(character))
        if ord(character) > 127:
            return False
    #fuera del for    
    return True
    
""" 
print(common_english('Docs To Go™ Free Office Suite'))
print(common_english('julian爱'))
print(common_english('Instagram')) 
"""
print('common_english_v1 => ',common_english('Instachat 😜'))

def common_english_v2(str_input):
    max_err = 0
    for character in str_input:
        #print(ord(character))
        if ord(character) > 127:
            max_err += 1  
    return max_err <= 3

print('common_english_v2 => ',common_english_v2('Instachat 😜'))
print('common_english_v2 => ',common_english_v2('Instachat 😜电视剧热播'))


common_english_v1 =>  False
common_english_v2 =>  True
common_english_v2 =>  False


# Filtrando solo las del idioma inglés
- Note que al final se imprimen los tamanios

In [14]:
def filter_english_only(data_list, name_index):
    cleaned_list = []
    for data in data_list:
        if common_english_v2(data[name_index]):
            cleaned_list.append(data)
    return cleaned_list

android_english_apps = filter_english_only(android_clean,0)
ios_english_apps = filter_english_only(apple_store_data[1:],1) 
print('apple_store_data => ',len(apple_store_data[1:]))
print('ios_english_apps => ',len(ios_english_apps))

print('android_clean => ', len(android_clean))
print('android_english_apps => ',len(android_english_apps))

apple_store_data =>  7197
ios_english_apps =>  6183
android_clean =>  9659
android_english_apps =>  9614


# Revome non-free apps

In [15]:
def remove_non_free(data_list, index):
    cleaned_list = []
    for data in data_list:
        price =  data[index].replace('$','')
        if float(price)== 0:
            cleaned_list.append(data)            
    return cleaned_list


ios_free_apps = remove_non_free(ios_english_apps,4) 
print('ios_english_apps => ',len(ios_english_apps))
print('ios_free_apps => ',len(ios_free_apps))

android_free_apps = remove_non_free(android_english_apps,7)
print('android_english_apps => ', len(android_english_apps))
print('android_free_apps => ',len(android_free_apps))


ios_english_apps =>  6183
ios_free_apps =>  3222
android_english_apps =>  9614
android_free_apps =>  8864


# Detectar mejores apps para invertir
La idea es que se detecten cuales son las apps favoritas por ambas plataformas, por ejemplo que se detecte un perfil de aplicación eg las de gamming. Lo anterior con el fin de:
- Saber si se hace una versión mínima para android y 6 meses despúes sacar la de ios, porque no tenemos dinero para ambos a la vez sin saber como reacciona el mercado.

Cómo hacer lo anterior? R/ Una idea puede ser mediante tablas de frecuencia :O

- Vamos a calcular las frecuencias de los generos que salen en el archivo 

In [16]:
def freq_table(data_list, index, percentage_format = True):
    freq_table = {}
    list_size = len(data_list)
    for data in data_list:
        key_column = data[index]
        if key_column in freq_table:
            freq_table[key_column] += (1/list_size)*100 if percentage_format else 1 
        else:
            freq_table[key_column] = (1/list_size)*100 if percentage_format else 1
    return freq_table

# Frequency by category 
#android_freq_table = freq_table(android_free_apps, 1)
#print(android_freq_table)

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        #ACA ARMAMOS UNA TUPLA DE TIPO (VAL, KEY)
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    #Usamos el sorted de listas porque no existe para dictionaries
    table_sorted = sorted(table_display, reverse = True)
    #Print de los dictionaries para que se ven mejor
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

"""



print('\n')
print('Android genres => ')
android_freq_table2 = display_table(android_free_apps, 9)
print(android_freq_table2)
print('\n')
"""

ios_freq_table_f = freq_table(ios_free_apps, 11, False)
print('IOS => ',len(ios_free_apps), sum(ios_freq_table_f.values()))
print(ios_freq_table_f)
ios_freq_table = display_table(ios_free_apps, 11)


#print(ios_freq_table_f)
#print(ios_freq_table)

IOS =>  3222 3222
{'Photo & Video': 160, 'Business': 17, 'Music': 66, 'Weather': 28, 'Medical': 6, 'Utilities': 81, 'Games': 1874, 'Entertainment': 254, 'Social Networking': 106, 'Catalogs': 4, 'News': 43, 'Sports': 69, 'Reference': 18, 'Travel': 40, 'Education': 118, 'Food & Drink': 26, 'Navigation': 6, 'Lifestyle': 51, 'Health & Fitness': 65, 'Book': 14, 'Finance': 36, 'Productivity': 56, 'Shopping': 84}
Games : 58.1626319056464
Entertainment : 7.883302296710134
Photo & Video : 4.965859714463075
Education : 3.6623215394165176
Social Networking : 3.2898820608317867
Shopping : 2.6070763500931133
Utilities : 2.5139664804469306
Sports : 2.1415270018621997
Music : 2.048417132216017
Health & Fitness : 2.0173805090006227
Productivity : 1.7380509000620747
Lifestyle : 1.5828677839851035
News : 1.3345747982619496
Travel : 1.2414649286157668
Finance : 1.1173184357541899
Weather : 0.8690254500310364
Food & Drink : 0.8069522036002481
Reference : 0.558659217877095
Business : 0.5276225946617009
Boo

# Preferencia de los usuarios por genero:

# IOS APPS

La idea es que se pueda detectar por cada genero que tan identificados los usuarios se sienten, para ello:
- En apps Ios se tiene la columna rating_count_tot que es el conteo de calificaciones por cada app
- En android sería por la cantidad de descargas

In [17]:
ios_by_genre = freq_table(ios_free_apps, 11, False)
#print(ios_by_genre)

def sort_dictionary(dictionary):
    table_display = []
    for key in dictionary:
        #ACA ARMAMOS UNA TUPLA DE TIPO (VAL, KEY)
        key_val_as_tuple = (dictionary[key], key)
        table_display.append(key_val_as_tuple)
        #Usamos el sorted de listas porque no existe para dictionaries
    table_sorted = sorted(table_display, reverse = True)
    #Print de los dictionaries para que se ven mejor
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

ios_user_ratings_by_genre = {}
for app in ios_free_apps:
    genre = app[11]
    rating_count_tot = float(app[5])
    if genre in ios_user_ratings_by_genre:
        ios_user_ratings_by_genre[genre] += rating_count_tot/ios_by_genre[genre]
    else:
        ios_user_ratings_by_genre[genre] = rating_count_tot/ios_by_genre[genre]
# ios_user_ratings_by_genre este dictionary guarda el promedio de 
# calificaiones por genero(tipo) de aplication
print(sort_dictionary(ios_user_ratings_by_genre))

Navigation : 86090.33333333333
Reference : 74942.11111111111
Social Networking : 71548.34905660375
Music : 57326.5303030303
Weather : 52279.89285714287
Book : 39758.5
Food & Drink : 33333.923076923085
Finance : 31467.944444444445
Photo & Video : 28441.543749999993
Travel : 28243.799999999992
Shopping : 26919.690476190473
Health & Fitness : 23298.01538461538
Sports : 23008.89855072465
Games : 22788.669690501574
News : 21248.023255813954
Productivity : 21028.41071428573
Utilities : 18684.45679012345
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661415
Business : 7491.117647058824
Education : 7003.983050847457
Catalogs : 4004.0
Medical : 612.0
None


 Dado el resultado anterior entonces Navigation es la que mas reviews o calificaciones tiene para IOS con un valor de 86090 y aunque representa el 0.18%
 de los registros del archivo eso no quiere decir que no es popular, quiere decir por ejemplo otras apps que se repiten en el archivo que genera ruido.

# Preferencia de los usuarios por genero:

# Android APPS

In [22]:
print('Android Category => ',len(android_free_apps))
android_freq_by_category = freq_table(android_free_apps, 1)
#print(android_freq_table1)

android_user_ratings_by_genre = {}
for app in android_free_apps:
    category = app[1]
    rating_count_tot = float(app[5].replace(',','').replace('+',''))
    if category in android_user_ratings_by_genre:
        android_user_ratings_by_genre[category] += rating_count_tot/android_freq_by_category[category]
    else:
        android_user_ratings_by_genre[category] = rating_count_tot/android_freq_by_category[category]


print(sort_dictionary(android_user_ratings_by_genre))

Android Category =>  8864
{'HOUSE_AND_HOME': 118027755.38410938, 'SHOPPING': 623748804.896482, 'PERSONALIZATION': 461059418.74938506, 'DATING': 75701115.51806082, 'EVENTS': 22473982.577777773, 'TRAVEL_AND_LOCAL': 1239548648.227245, 'ART_AND_DESIGN': 176068742.1754385, 'PHOTOGRAPHY': 1581347386.0597694, 'PRODUCTIVITY': 1488029050.414371, 'MEDICAL': 10685606.939808294, 'BUSINESS': 151777398.66732082, 'GAME': 1381721703.0718832, 'LIFESTYLE': 127448034.06520125, 'BOOKS_AND_REFERENCE': 777178846.3494719, 'COMMUNICATION': 3408750402.9848027, 'SPORTS': 322529062.26285523, 'TOOLS': 957435324.713795, 'FINANCE': 123005061.03804839, 'EDUCATION': 162521009.70873788, 'ENTERTAINMENT': 1031832169.4117625, 'FOOD_AND_DRINK': 170622935.3512727, 'WEATHER': 449802456.51830906, 'BEAUTY': 45485783.24528301, 'PARENTING': 48096384.93793105, 'NEWS_AND_MAGAZINES': 846439179.3806437, 'LIBRARIES_AND_DEMO': 56596971.06506023, 'FAMILY': 327581690.90788025, 'AUTO_AND_VEHICLES': 57378251.305365786, 'HEALTH_AND_FITNES